src/core/net/mdns.cpp

```diff
@@ -0,0 +1,3887 @@
+// Core::FireTime
+
+void Core::FireTime::SetFireTime(TimeMilli aFireTime)
+{
+    if (mHasFireTime)
+    {
+        VerifyOrExit(aFireTime < mFireTime);
+    }
+
+    mFireTime    = aFireTime;
+    mHasFireTime = true;
+
+exit:
+    return;
+}
+
+void Core::FireTime::ScheduleFireTimeOn(TimerMilli &aTimer)
+{
+    if (mHasFireTime)
+    {
+        aTimer.FireAtIfEarlier(mFireTime);
+    }
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+// Core::Entry
+
+Core::Entry::Entry(void)
+    : mState(kProbing)
+    , mProbeCount(0)
+    , mMulticastNsecPending(false)
+    , mUnicastNsecPending(false)
+    , mAppendedNsec(false)
+{
+}
+
+void Core::Entry::Init(Instance &aInstance)
+{
+    // Initializes a newly allocated entry (host or service)
+    // and starts it in `kProbing` state.
+
+    InstanceLocatorInit::Init(aInstance);
+    StartProbing();
+}
+
+void Core::Entry::SetState(State aState)
+{
+    mState = aState;
+    ScheduleCallbackTask();
+}
+
+void Core::Entry::Register(const Key &aKey, const Callback &aCallback)
+{
+    if (GetState() == kRemoving)
+    {
+        StartProbing();
+    }
+
+    mKeyRecord.UpdateTtl(DetermineTtl(aKey.mTtl, kDefaultKeyTtl));
+    mKeyRecord.UpdateProperty(mKeyData, aKey.mKeyData, aKey.mKeyDataLength);
+
+    mKeyCallback = aCallback;
+    ScheduleCallbackTask();
+}
+
+void Core::Entry::Unregister(const Key &aKey)
+{
+    OT_UNUSED_VARIABLE(aKey);
+
+    VerifyOrExit(mKeyRecord.IsPresent());
+
+    mKeyCallback.Clear();
+
+    switch (GetState())
+    {
+    case kRegistered:
+        mKeyRecord.UpdateTtl(0);
+        break;
+
+    case kProbing:
+    case kConflict:
+        ClearKey();
+        break;
+
+    case kRemoving:
+        break;
+    }
+
+exit:
+    return;
+}
+
+void Core::Entry::ClearKey(void)
+{
+    mKeyRecord.Clear();
+    mKeyData.Free();
+}
+
+void Core::Entry::SetCallback(const Callback &aCallback)
+{
+    mCallback = aCallback;
+    ScheduleCallbackTask();
+}
+
+void Core::Entry::ScheduleCallbackTask(void)
+{
+    switch (GetState())
+    {
+    case kRegistered:
+    case kConflict:
+        VerifyOrExit(!mCallback.IsEmpty() || !mKeyCallback.IsEmpty());
+        Get<Core>().mEntryTask.Post();
+        break;
+
+    case kProbing:
+    case kRemoving:
+        break;
+    }
+
+exit:
+    return;
+}
+
+void Core::Entry::InvokeCallbacks(void)
+{
+    Error error = kErrorNone;
+
+    switch (GetState())
+    {
+    case kConflict:
+        error = kErrorDuplicated;
+        OT_FALL_THROUGH;
+
+    case kRegistered:
+        mKeyCallback.InvokeAndClear(GetInstance(), error);
+        mCallback.InvokeAndClear(GetInstance(), error);
+        break;
+
+    case kProbing:
+    case kRemoving:
+        break;
+    }
+}
+
+void Core::Entry::StartProbing(void)
+{
+    SetState(kProbing);
+    mProbeCount = 0;
+    SetFireTime(TimerMilli::GetNow() + kInitalProbeDelay);
```

Comment:
RFC 6762 8.1 Probing - When ready to send its Multicast DNS probe packet(s) the host should
   first wait for a short random delay time, uniformly distributed in
   the range 0-250 ms.  This random delay is to guard against the case
   where several devices are powered on simultaneously, or several
   devices are connected to an Ethernet hub, which is then powered on,
   or some other external event happens that might cause a group of
   hosts to all send synchronized probes.